src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,361 @@
+                        : Dns.GetHostAddresses(host);
+
+                    if (addresses.Length == 0)
+                    {
+                        throw new Exception("No suitable IPv4 address.");
+                    }
+
+                    ipv4Address = addresses[0];
+                }
+
+                if (ipv4Address == null)
+                {
+                    buffer[4] = 0;
+                    buffer[5] = 0;
+                    buffer[6] = 0;
+                    buffer[7] = 255;
+                }
+                else
+                {
+                    ipv4Address.TryWriteBytes(buffer.AsSpan(4), out int bytesWritten);
+                    Debug.Assert(bytesWritten == 4);
+                    if (buffer[4] == 0 && buffer[5] == 0 && buffer[6] == 0)
+                    {
+                        // Invalid IP address used by SOCKS4a to represent remote DNS.
+                        // In case we don't have a domain name, throwing.
+                        throw new Exception("Invalid ip address.");
+                    }
+                }
+
+                int uLen = Encoding.UTF8.GetBytes(username, buffer.AsSpan(8));
+                buffer[8 + uLen] = 0;
+                int totalLength = 9 + uLen;
+
+                if (isVersion4a && ipv4Address == null)
+                {
+                    // https://www.openssh.com/txt/socks4a.protocol
+                    int aLen = Encoding.UTF8.GetBytes(host, buffer.AsSpan(9 + uLen));
+                    buffer[9 + uLen + aLen] = 0;
+                    totalLength += aLen + 1;
+                }
+
+                await WriteAsync(stream, buffer.AsMemory(0, totalLength), async, cancellationToken).ConfigureAwait(false);
+
+                // +----+----+----+----+----+----+----+----+
+                // | VN | CD | DSTPORT |      DSTIP        |
+                // +----+----+----+----+----+----+----+----+
+                //    1    1      2              4
+                await ReadToFillAsync(stream, buffer.AsMemory(0, 8), async, cancellationToken).ConfigureAwait(false);
+                if (buffer[0] != ProtocolVersion4)
+                {
+                    throw new Exception("Bad protocol version");
+                }
+                if (buffer[1] != CD_SUCCESS)
+                {
+                    throw new Exception("Connection failed");
+                }
+                // response address not used
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+        }
+
+        private static async ValueTask WriteAsync(Stream stream, Memory<byte> buffer, bool async, CancellationToken cancellationToken)
+        {
+            if (async)
+            {
+                await stream.WriteAsync(buffer, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                stream.Write(buffer.Span);
+            }
+        }
```

Comment:
```suggestion
        private static ValueTask WriteAsync(Stream stream, Memory<byte> buffer, bool async)
        {
            if (async)
            {
                return stream.WriteAsync(buffer);
            }
            else
            {
                stream.Write(buffer.Span);
                return default;
            }
        }
```

We don't have to pass the CT here either since we're registered for the stream.

Reply:
So the CT is just used for DNS?
And how about DNS in sync path?

Reply:
Since we're registered to the CT to dispose the underlying stream, we don't have to pass in the CT to stream reads/writes.
We still have to cancel the DNS calls. For the sync path, I don't think we can do anything here.